perl/lib/Lingy/HashMap.pm

```diff
@@ -6,15 +6,26 @@ use base 'immutable::map', 'Lingy::Class';
 use Lingy::Common;
 
 sub new {
-    my ($class, $list) = @_;
+    my ($class, $data) = @_;
     my (@keys, %vals);
-    for (my $i = @$list - 2; $i >= 0; $i -= 2) {
-        my $key = $class->_get_key($list->[$i]);
-        if (not exists $vals{$key}) {
-            unshift @keys, $key;
-            $vals{$key} = $list->[$i + 1];
+
+    if (ref($data) eq 'HASH') {
+        @keys = keys %$data;
```

Comment:
I would `= sort keys %$data` here to be deterministic, since Perl hash key order is never deterministic.